fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1839,7 +1850,54 @@ private boolean canDeleteWhereForIndexOnStoredTypes(@Nonnull final IndexMaintain
             }
         }
 
-        private boolean canDeleteWhereForIndexOnSyntheticTypes(@Nonnull final IndexMaintainer indexMaintainer) {
+        private boolean canDeleteWhereForIndexOnJoinedTypes(@Nonnull final IndexMaintainer indexMaintainer) {
+            final Index index = indexMaintainer.state.index;
+            final Collection<RecordType> recordTypesForIndex = recordMetaData.recordTypesForIndex(index);
+
+            if (Key.Expressions.hasRecordTypePrefix(index.getRootExpression())) {
+                // If the index has a record type prefix, we need to reject the deletion, as the referenced
+                // record type key is different for the synthetic type and the requested type
+                return false;
+            }
+            JoinedRecordType joinedRecordType = null;
+            for (RecordType indexRecordType : recordTypesForIndex) {
+                // currently do not support multitype indexes with joined types
+                if (!(indexRecordType instanceof JoinedRecordType) || joinedRecordType != null) {
+                    // this should only be called when all of them are JoinedRecordTypes
+                    return false;
+                }
+                joinedRecordType = (JoinedRecordType)indexRecordType;
+            }
+
+            if (typelessComponent != null &&
```

Comment:
Should there be a check that `recordTypeKeyComparison == null` here? I think it may technically be unnecessary, though I think we'd want slightly different logic for the single-type case. In particular:

1. We can clear out the index entirely if `typelessComponent` is null and there are no outer joins
2. We don't actually need to find a join that matches, but if there isn't one, we just need to match against the constituent of the relevant type
3. If we find a join on the relevant type that matches the delete-where criteria, we could theoretically translate that to a query on a different constituent (potentially applying some compensation to account for different field names)

I think what we have here is sort of a special-case of (3), in that we require that the non-participating types match the predicate even if that's not necessary.